compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/ClassReference.kt

```diff
@@ -395,6 +395,26 @@ public fun ClassReference.allSuperTypeClassReferences(
     .distinct()
 }
 
+/**
+ * Resolves the concrete type for a super type's generic types.
+ *
+ * e.g. this will find [Int] when passing [parameterIndex] of 1 in the following example:
+ * ```
+ * interface SuperClass<T, R>
+ *
+ * interface ImplementingClass : SuperClass<String, Int>
+ * ```
+ *
+ * @param parameterIndex: The index of the super type parameter to resolve.
+ */
+@ExperimentalAnvilApi
+public fun Psi.resolveSuperTypeGenericTypeReference(
```

Comment:
If we decide to move forward, then we should make the existing internal function public and not add a new extension function.

Reply:
> That's the problem why we didn't expose this before. It only works for Psi, but not descriptors and this makes me very hesitant to merge this. Notice the comment here:

Thanks for taking a look Ralf, I didn't notice that comment. So to clarify, was the previous conclusion that resolving the type for this use-case would not be possible for descriptors? Or just that we haven't looked into it enough yet? I did not look closely at descriptor options while making the changes for this PR since it was already PSI-only.

Reply:
There are solutions to make it work for Descriptors, but the process is tedious and you need a lot of context where the generic type is used in order to resolve concrete argument. Unlike Psi Descriptors don't tell you what the type arguments are. For example
```
interface A<T> {
  val a: T
}

class B : A<String> {
  override val a: String
}
```
To know what concrete type the property `a` for class `B` is you'd take a look at the declared super types with Psi. Easy.

For Descriptors you can actually look at `B.a` directly and it'll give you `String`. But there's no easy connection between the generic type `T` and it being resolved to `String`. In other words you cannot easily ask what `T` is bound to. 

A generic solution could be to find a property / function parameter / function return type /.. where `T` is used within class `A` and then get the actual type from class `B`. That would work, but is a lot of parsing and checking to find the usage of `T`.

What Anvil itself did internally is it knew that it needed a specific return type for a function, so it directly resolved the function and its return type. There's a similar approach for parameters. But we never implemented a generic way, because we didn't need to.

Reply:
@vRallev We currently use ClassDescriptor directly to resolve super type arguments: `classDescriptor.typeConstructor.supertypes.single { ... }.arguments.single().type.classDescriptor()`.

For context, we have this structure (simplified):
```kotlin
interface FragmentFactory<Arg, Result> {
    operator fun invoke(arg: Arg): FragmentWithResult<Result>
}

interface FragmentFactoryContributor<T : FragmentFactory<*, *>>

interface SomeFragmentFactory : FragmentFactory<String, String>

class SomeFragment : Fragment(), FragmentFactoryContributor<SomeFragmentFactory>
```

And our code generator creates an implementation of SomeFragmentFactory. So we need to resolve SomeFragmentFactory, as well as Arg and Result types.

I tried using anvil from this branch, and was able to resolve type argument of FragmentFactoryContributor using the new method, since SomeFragment's ClassReference is a Psi reference coming directly from `projectFiles.classAndInnerClassReferences(module)`, but the resulting reference to SomeFragmentFactory is a Descriptor reference, and I can't use it to resolve Arg and Result types.

We are able, however, to resolve everything using ClassDescriptor directly:
```kotlin
val factoryType = fragmentClassDescriptor.typeConstructor.supertypes.single {
    it.classDescriptor().asClassName() == fragmentFactoryContributorClassName
}.arguments.single().type
val factorySuperType = factoryType.constructor.supertypes.single()
val argumentType = factorySuperType.arguments[0].type.classDescriptor().asClassName()
val resultType = factorySuperType.arguments[1].type.classDescriptor().asClassName()
```

Reply:
Oh my gosh, ignore what I said above. This is the API I needed and wasn't able to find for such a long time:
```kotlin
descriptor.typeConstructor.supertypes.single().arguments
```
That's something we can work with.

Reply:
Thanks for the input @romankivalin. The new fix is here: https://github.com/square/anvil/pull/586 We plan to merge this soon. Maybe you can give the latest snapshot a try after that and report back in the issue if it resolves to request.

Reply:
@vRallev Thanks, I was able to rewrite our code generator using only ClassReference and TypeReference apis on snapshot version.